[XLA] vectorize row reduction for even row size

nouiz opened the pull request:
This depend on LLVM commit: https://reviews.llvm.org/D74444 that is recently merged in XLA.

This PR vectorize row reduction for even row size.
For odd row size, every other row is unaligned. So would need a more complicated fix.
On a reduction of size 32x131070, I have no speed up for a max reduction, but I have 1.21x speed on the sum+pw reduction from the softmax in float16.

This was benchmarked on the row major "Softmax" TF operations.
The optimized graph contain 3 main operations: reduce_8 (the max reduction), fusion_1 (the exponential and sum reduction), fusion (the post normalization pw operations, ignored).
This was benchmarked with a matrix of input shape (32x131070) with float16 dtype:

| Kernel     | Before PR | After PR | Speed up |
| ---------- | ---------- | ---------- | ---------- |
| fusion_1  | 39.827us  | 33.843us | 1.18x |
| reduce_8 | 22.112us  | 22.047us | 1x |

Here is the not optimized HLO:
```
HloModule cluster_0
%max_ {
  %x = f16[] parameter(0)
  %y = f16[] parameter(1)
  ROOT %maximum = f16[] maximum(%x, %y)
}

%add_ {
  %x = f32[] parameter(0)
  %y = f32[] parameter(1)
  ROOT %add = f32[] add(%x, %y)
}

ENTRY %main {
  %arg0.1 = f16[32,131070]{1,0} parameter(0)
  %constant.3 = f16[] constant(-inf)
  %pad1 = f16[32,131070] pad(%arg0.1, %constant.3), padding=0_0_0x0_0_0
  %reduce.8 = f16[32]{0} reduce(%pad1, %constant.3), dimensions={1}, to_apply=%max_
  %broadcast.9 = f16[32,131070]{1,0} broadcast(%reduce.8), dimensions={0}
  %subtract.131070 = f16[32,131070]{1,0} subtract(%arg0.1, %broadcast.9)
  %exponential.11 = f16[32,131070]{1,0} exponential(%subtract.131070)
  %convert.12 = f32[32,131070]{1,0} convert(%exponential.11)
  %constant.13 = f32[] constant(0)
  %pad2 = f32[32,131070] pad(%convert.12, %constant.13), padding=0_0_0x0_0_0
  %reduce.18 = f32[32]{0} reduce(%pad2, %constant.13), dimensions={1}, to_apply=%add_
  %convert.19 = f16[32]{0} convert(%reduce.18)
  %broadcast.20 = f16[32,131070]{1,0} broadcast(%convert.19), dimensions={0}
  ROOT %divide.21 = f16[32,131070]{1,0} divide(%exponential.11, f16[32,131070]{1,0} %broadcast.20)
}
```

nouiz committed:
Add env variable to allow different reduction case.

nouiz committed:
Enable the new vectorized+strided iteration mode only for row reduction.

nouiz committed:
Move the vector size into KernelMappingScheme

nouiz committed:
If some blocks fit completly, unroll to vectorize for them.

nouiz committed:
Only vectorize when the tile completly fit.

nouiz committed:
Limit default vectorization to we know LLVM won't vectorize

nouiz committed:
Rename a variable

nouiz committed:
Add comment

nouiz committed:
Enable vectorization for even row size by default.

nouiz committed:
Remove not working code path for odd row size.

nouiz committed:
Add test

nouiz committed:
Rename a variable

nouiz committed:
Code simplification

nouiz committed:
Remove useless indentation

nouiz committed:
Remove print

nouiz committed:
Fix new latent bug.

nouiz committed:
Better test

nouiz committed:
NFC: clang-format

nouiz committed:
Better comment

nouiz commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -76,19 +76,35 @@ namespace gpu {
 class KernelMappingScheme {
  public:
   enum { DimZ = 0, DimY, DimX, DimTot };
+  // TODO: rename Dilated to Strided?
```
The current XLA code is using dilated. I think more people would better understand it if we rename it to Strided.
If you agree, I'll update the code.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_if, int64 max_element, int64 vector_size) {
+          for (int64 j = 0; j < x_num_steps / vector_size; j++) {
+            // Prep some values. If we do not do this, LLVM doesn't vectorize.
```
Could you expand on what exactly is being done here?

nouiz replied:
This isn't needed anymore, so I'll remove this.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+              int old_j = j * vector_size + i;
+              llvm::Value* x_loc = x_loc_base;
+              IrArray::Index source_idx_x = source_idx_x_base;
+              if (i > 0) {
```
Is the branch worth it here? LLVM will optimize away the addition with zero.

nouiz replied:
This isn't needed anymore, so I'll remove this.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+                return emit_elem_function(source_idx_x, y_loc, x_loc, old_j);
+              };
+              if (add_if) {
+                ksl->If(loop_name + "_x_in_tile",
```
Is this check worth it? Do the benchmark shows that the performance has a non-negligible drop if we don't always check for tiling conditions?

nouiz replied:
This one make a difference no. For the even row size, LLVM isn't able to remove the condition.
So LLVM doesn't vectorize at all without this.
I just disabled this path and I confirm this is needed for the even row size. Otherwise one of the new test fail.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+          // For odd row size every other row isn't aligned, so can't be
+          // vectorized.
+          ksl->If(loop_name + "_is_full_tile",
+                  // if (block fully fit) {fast path} else {slow path}
```
Do we actually need this fast_path/slow_path divergence? I understand the optimization, but previously I have removed a lot of those, because they add a lot of code complexity, but often lead to negligible increases in performance.

nouiz replied:
This is needed. See my answer to comment: https://github.com/tensorflow/tensorflow/pull/37260#discussion_r387805494

cheshire replied:
Sorry I'm still a bit confused there, are you saying that when `add_index_boundary_conditions` is `false`, vectorization does not happen? Do we even need unrolling in that case then?

nouiz replied:
For LLVM to vectorize even row size, we need 2 things to have the vectorized loads and handle the partial tile at the end of the row.
- One path without the condition between each loads so that LLVM vectorize them.
- One path with the condition to manage for the last warp that loads only part of the elements.

The add_index_boundary_conditions when false, won't add the condition between loads. So LLVM can vectorize the loads. But when true, it handle the partial tile at the end of the row.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+                  // if (block fully fit) {fast path} else {slow path}
+                  // tile_width is always exact. For the last block,
+                  // it will be the exact number of elements left.
+                  b_.CreateICmpEQ(constant(mapping_scheme.GetTileSizeFor(2)),
```
What's `2` here?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1860,12 +1860,14 @@ namespace {
 // translated to complex LLVM IR, such as loops, and prevent vectorization.
-bool MayPreventVectorization(const HloInstruction& hlo) {
+bool MayPreventVectorization(const HloInstruction& hlo,
+                             const bool tolerate_reduce = false) {
```
Also could you expand on why do we need this parameter? In general, I'd like to avoid having boolean parameters as much as possible, as they greatly increase the code complexity by doubling the number of possible execution paths.

nouiz replied:
I realized that it isn't needed. So I remove it.

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -76,19 +76,35 @@ namespace gpu {
  public:
   enum { DimZ = 0, DimY, DimX, DimTot };
+  // TODO: rename Dilated to Strided?
+  // LinearIndexing mean each thread reads consecutive elements.
```
"means", "reads", "takes", "conserves", probably more readable if the comment for each enum value is on top of it.

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -133,12 +150,16 @@ class KernelMappingScheme {
+  // vector_size_ only supported for row reduction.
+  // Must be a divisor of tile_sizes_[2]/num_threads_x.
+  // Interesting values are 2 and 4 to trigger vectorized loads on GPUs
+  // While keeping memory coalescing.
```
"while"

cheshire commented on tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc:
```diff
@@ -228,13 +228,14 @@ bool MaybeLoadPtxFromFile(const HloModule* module, std::string* ptx) {
   std::string prefix = xla::FilenameFor(*module, "", *ptx);
   std::string matched_filename;
-  for (const string filename :
+  for (const string full_filename :
```
If this change is needed, it should be a separate PR.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+                          ParseAndReturnVerifiedModule(hlo_text));
+  CompileAndOptionallyVerifyPtx(std::move(optimized_module),
+                                R"(
+CHECK: ld.global.nc.v2.f32
```
In general, do you think this is better tested on ptxas level or an LLVM level or both?

cheshire replied:
If we test on the LLVM level, the test might also "just work" for AMD (though I understand that having this conversation is ironical)

nouiz replied:
There is a conceptual problem here. Some of the vectorization are done in the NVPTX/AMD back-end, not inside LLVM optimization. This is related to the LLVM changes I did that was in the NVPTX back-end.
So I do not see how this can be tested at the LLVM IR level without a big overhaul of LLVM.

cheshire replied:
OK makes sense.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+                          ParseAndReturnVerifiedModule(hlo_text));
+  CompileAndOptionallyVerifyPtx(std::move(optimized_module),
+                                R"(
+// TODO: Make this a vectorized load
```
What's the TODO item here? Could you expand on what is not working and why?

nouiz replied:
I didn't fully investigate this yet. But there is problem in the optimization phase in LLVM.
My expectation is that the branch optimization are done before the vectorization optimization (as they are in the NVPTX/AMD back-end). Those optimization interact together. As both branch are similar, I think the end of both branch end up being the same. I suppose LLVM hoist that out of the branch. Then we do not have both loads in the same basic block. So they aren't optimized.

This is my current not validate hypothesis.

But this already vectorize 3 out of 4 loads and give speed up in my benchmark. So I think it is good to split all the changes and merge now this part that is working.

nouiz replied:
I do not to fix this TODO in this PR. Do you have an idea or a comment that would be more clear for you? Just "TODO: make the missing vectorization happen"?

cheshire replied:
OK I understand, usually we have a tracking bug or the googler owning each TODO, but I understand neither makes too much sense here yet.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+CHECK: ld.global.nc.v2.f32
+CHECK: ld.global.nc.v2.f32
+CHECK-NOT: ld.global.nc.v2.f32
+// TODO: find how to modify LLVM/opt to get this merged? vectorize before some loop opt?
```
Again, to me (and someone unfamiliar with the code) this comment is cryptic.

cheshire replied:
Not replied yet.

nouiz replied:
I changed it in the new commit.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+
+namespace {
+
+class ReductionVectorizationTest : public GpuCodegenTest {
```
I'm curious what happens in the input fusion case.

nouiz committed:
Use constant to make the code clearer.

nouiz committed:
Small NFC from review.

nouiz committed:
Use strided instead of dilated. This is more know in our community working on the software.

nouiz committed:
NFC: Rename a variable

nouiz committed:
Simplify the code. The condition is only used in one of the path, so no need to test it.

nouiz committed:
Remove a condition that is optimized by LLVM and doesn't remove vectorization.

nouiz committed:
Simplify code. This isn't needed anymore.

nouiz committed:
Simplify comment.

nouiz commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
+      /*step=*/constant(1), [&](llvm::Value* y_indvar) {
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_if, int64 max_element, int64 vector_size) {
```
I updated the parameters names and removed one that wasn't used.
The code was also simplified, so I think we can let it as a lambda.
If you prefer a separate function, just tell me.

nouiz commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,7 +3170,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
-  if (!dilated_x && !reduction_dimensions.is_row_reduction) {
+  KernelMappingScheme::IndexingOrder indexing_order;
+  if (reduction_dimensions.is_row_reduction &&
+      // Only try to vectorize+coales memory access for row of even size.
```
I tried adding an kPad node in the XLA graph. If you suggest a different padding, tell me. This have a few problems:
- This cause the fused reduction with exponential for softmax to be 2 fusion instead of one. 
   - The fusion input shape and reduction input shape aren't the same and the fusion optimizer reject that case.
   - I hacked a fix the fusion optimization, but this also cause problem in the generated code where the shapes isn't the right one.
- The final PTX code still ends up with some conditions in it. But the conditions break the vectorization for each some of the loads for all threads. So the code should be less optimized. I didn't benchmark if this make a difference, but the current partially optimized even case do not have all the speed up I was expecting if it was fully vectorized.

I have 2 ideas that could help odd row shapes. This could remove this need. But this would delay much this PR that is fully working and useful. Also it isn't even sure they will works. So I think it is useful to get this change in now.

nouiz commented:
I did the change or answer all comments.

Just a note that for odd row size, this is also complicated. There is memory order optimization problem in LLVM. I already fixed one, but the problem left is probably more complicated. I'm still trying a few different options before trying to change LLVM optimization passes.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1863,9 +1868,10 @@ namespace {
-                          [](const HloInstruction* instr) {
+                          [&](const HloInstruction* instr) {
                             switch (instr->opcode()) {
                               case HloOpcode::kReduce:
```
Since `false` is default, just removing it from the list would suffice?

cheshire replied:
Also could you add a test case affected by this particular the change in this particular function? I'm not sure what would be a good testing boundary, maybe LLVM?

nouiz replied:
You are right that I could remove it. But as I want to explicitly to return false, I moved it the case to just before the default case. I see this as having a whitelist and a blacklist and a default. Even if the default is the same as the black list, I prefer to have it explicit. What do you think of the new version?

As the behavior is the same as the default value, I'm not sure how to test this. Any idea?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1892,6 +1898,8 @@ bool MayPreventVectorization(const HloInstruction& hlo) {
         return false;
     }
+  } else if (hlo.opcode() == HloOpcode::kReduce) {
+    return false;
```
Do we need to be specific about what reduction is vectorized?

nouiz replied:
I suppose you mean to check what inside the to_apply() HloComputation of the reduction.
In theory yes. But it wasn't done before. I'm not sure it is worth doing it.
So I'll add a todo about this to let people know about this.

cheshire replied:
Maybe let's make it more specific: state what kind of reductions still prevent vectorization.

cheshire replied:
> Reduction doesn't prevent vectorization. This is the point

I understand, I mean: could you specify in the comment explicitly what needs to be checked and why, e.g. "return false only for row reductions which are ..."?

nouiz replied:
This function is independent of the reduction being a row or a columns reductions. I updated it to be tell what to check.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,8 +1928,12 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+    return b->CreateMul(thread_id_x,
+                        llvm::ConstantInt::get(index_ty, vector_size));
   }
   int64 x_num_steps =
```
Perhaps insert a CHECK on what the indexing order has to be in the third case?

cheshire replied:
I mean, you are assuming that the indexing order has to be linear, so it's useful to check for that.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,8 +1928,12 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+  if (mapping_scheme.GetIndexingOrder() == kStridedIndexingX) {
     return thread_id_x;
+  } else if (mapping_scheme.GetIndexingOrder() == kLinearStridedIndexingX) {
+    int vector_size = mapping_scheme.GetVectorSize();
```
Nit: I would extract a lambda for creating constants, since this is used twice here already, and then inline the `vector_size` in the call.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,64 @@ void IrEmitterUnnested::EmitTile(
-               }
-             }
-           });
+  int vector_size = mapping_scheme.GetVectorSize();
```
Maybe `int64` as well for consistency, and group the declaration together with `step_x`

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,64 @@ void IrEmitterUnnested::EmitTile(
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_index_boundary_condition,
+                          int64 vector_size) {
```
Why not just take `vector_size` from the captured environment?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1991,64 @@ void IrEmitterUnnested::EmitTile(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_index_boundary_condition,
+                          int64 vector_size) {
+          for (int64 j = 0; j < x_num_steps / vector_size; j++) {
```
For consistency, let's have both `i` and `j` as `int`?

cheshire replied:
Also is it possible to have more descriptive names? In particular, for `old_j`?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,7 +3163,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
                                             reduction_dimensions.dimensions[2]);
 
-  if (!dilated_x && !reduction_dimensions.is_row_reduction) {
+  KernelMappingScheme::IndexingOrder indexing_order;
```
Use the lambda trick:
```
IndexingOrder indexing_order = [&] {
  ...
  return kLinearStridedIndexingX;
}();
```

nouiz replied:
I did. Normally I use it for const variable with complex initialization. What gain do you see here?
I modify that variable later in some condition.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,7 +3163,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+
+  if (indexing_order == kLinearIndexingX &&
+      !reduction_dimensions.is_row_reduction) {
     // Vectorized loads: a single thread reduces two adjacent columns.
```
But not all row reductions are vectorized?

nouiz replied:
I'm not sure I understand your question.
This comment was already in the code in the past.
This path is for column reduction, where each thread reduce 2 consecutive output value at the same time.

row reduction aren't always vectorized. When the tile fit, they are always vectorized. For even row size, it is partially vectorized. The missing vectorized is due to a problem of LLVM optimization pass order.
For odd row size, I do not trigger the new indexing order as it doesn't get vectorized by LLVM and this slow down my benchmark.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3133,11 +3196,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+  int vector_size = 1;
+  if (indexing_order == kLinearStridedIndexingX) {
+    if (reduction_dimensions.dimensions[2] % 2 == 0 &&
+        // As XLA unroll and suppose LLVM will vectorize,
```
"Assuming XLA will perform the unrolling and LLVM will vectorize" ?

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -76,19 +76,33 @@ namespace gpu {
  public:
   enum { DimZ = 0, DimY, DimX, DimTot };
+  enum IndexingOrder {
+  // Threads read consecutive elements.
```
Commend aligned with the variable name.

cheshire replied:
Also I think it's more readable to write `Thread reads consecutive elements`, here and below, even though there are multiple threads.

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -76,19 +76,33 @@ namespace gpu {
+  // Threads read a few consecutive elements then take a strided
+  // step. This can trigger vectorized reads and keep memory
+  // coalescing.
+    LinearStridedIndexingX
```
Maybe `UnrolledStridedIndexingX`?

nouiz replied:
Personnally, I prefer the name I pick. But if you prefer your, I can change it.
I do not like using Unrolled as the code is unrolled for the 3 case. So we should put it in all names or no names.
As this is a mix  of Linear and Strided indexing, I like putting the 2 names to make it clear.

But I could change the order of the name to StridedLinearIndexingX. The Linear part is inside, so the part name order would match the loop order.

It could also be LinearVectorizedIndexingX, but as the vectorization isn't done by XLA, I still perfer StridedLinearIndexingX.

Other idea? What do you think of the new name suggestion?

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -133,12 +148,16 @@ class KernelMappingScheme {
+  // elements in the X dimension of a tile, each threads process
+  // n=tile_size_x/num_threads_x elements.
+  // indexing_order_ define which tile's elements each thread reads.
+  const IndexingOrder indexing_order_;
```
Newline separating the members.

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -133,12 +148,16 @@ class KernelMappingScheme {
+  // n=tile_size_x/num_threads_x elements.
+  // indexing_order_ define which tile's elements each thread reads.
+  const IndexingOrder indexing_order_;
+  // vector_size_ only supported for row reduction.
```
"Only supported for row reduction, ..."

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+  EXPECT_TRUE(RunAndCompare(hlo_text, ErrorSpec{1e-5, 1e-5}));
+}
+
+TEST_F(ReductionVectorizationTest, EvenColums) {
```
`Columns`

nouiz committed:
NFC: better code/comments following code review.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1874,6 +1878,8 @@ bool MayPreventVectorization(const HloInstruction& hlo) {
                               case HloOpcode::kPower:
                               case HloOpcode::kAtan2:
                                 return true;
+                              case HloOpcode::kReduce:
```
I would still just remove these two lines

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1993,59 @@ void IrEmitterUnnested::EmitTile(
+      /*step=*/constant(1), [&](llvm::Value* y_indvar) {
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_index_boundary_condition) {
```
Please refactor into a separate functino.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +1993,59 @@ void IrEmitterUnnested::EmitTile(
+
+        if (!x_tile_fits &&
+            mapping_scheme.GetIndexingOrder() == kLinearStridedIndexingX) {
+          // Only try this path when we try to vectorize the loads.
```
The comment reads better above the `if`

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3133,11 +3193,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+    if (reduction_dimensions.dimensions[2] % 2 == 0 &&
+        // Assuming XLA will perform the unrolling and LLVM will vectorize,
+        // disable the unroll for case that LLVM doesn't vectorize.
+        !MayPreventVectorization(*unnested_hlo)) {
```
Is this why the change to `MayPreventVectorization` was necessary? Or is that for fused computations as well?

nouiz committed:
Comment and reuse useless case.

nouiz committed:
Move a lambda to its own function.

nouiz committed:
Add a CHECK.

nouiz reviewed (commented):
I replied or did the change for all comments.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1927,50 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+  return b->CreateMul(thread_id_x, constant(x_num_steps));
+}
+
+auto Unroll(bool add_index_boundary_condition, int64 x_num_steps, int64 step_x,
```
static void?

cheshire replied:
+ a short docstring on what this function is doing

cheshire replied:
+ Also perhaps a better name would be `check_x_tile_bounds`?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1927,50 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+              .AddOffsetToDim(constant(j * step_x * vector_size + i), kDimX,
+                              &b_);
+      auto emit_element = [&] {
+        return (*emit_elem_function)(source_idx_x, y_loc, x_loc, linear_index);
```
You don't need to explicitly dereference the function pointer, `emit_elem_function(...)` will do.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1987,21 +2033,28 @@ void IrEmitterUnnested::EmitTile(
+                             tile_width, source_idx, b_, &emit_elem_function);
+             };
+
+             // Only try this path when we try to vectorize the loads.
```
"when we are vectorizing the loads" I think it more clear, we are not merely trying.

nouiz replied:
I consider that we try to vectorize as it is LLVM that does the vectorization, XLA only unroll.
I'll rephrase it.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1987,21 +2033,28 @@ void IrEmitterUnnested::EmitTile(
+             };
+
+             // Only try this path when we try to vectorize the loads.
+             // Special case when the tile doesn't fit completly for even row
```
"completely"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1987,21 +2033,28 @@ void IrEmitterUnnested::EmitTile(
+
+             // Only try this path when we try to vectorize the loads.
+             // Special case when the tile doesn't fit completly for even row
+             // size. For odd row size every other row isn't aligned, so can't
```
"so it can't be vectorized by LLVM"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1987,21 +2033,28 @@ void IrEmitterUnnested::EmitTile(
+             if (!x_tile_fits &&
+                 mapping_scheme.GetIndexingOrder() == kLinearStridedIndexingX) {
+               ksl->If(loop_name + "_is_full_tile",
+                       // if (block fully fit) {fast path} else {slow path}
```
I would remove this line entirely, I think "fast path"/"slow path" is somewhat misleading, as we don't get vectorization at all on the slow path.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1987,21 +2033,28 @@ void IrEmitterUnnested::EmitTile(
+                 mapping_scheme.GetIndexingOrder() == kLinearStridedIndexingX) {
+               ksl->If(loop_name + "_is_full_tile",
+                       // if (block fully fit) {fast path} else {slow path}
+                       // tile_width is always exact. For the last block,
```
I don't understand what does it mean to be always "exact"?

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,7 +3171,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
-  if (!dilated_x && !reduction_dimensions.is_row_reduction) {
+  KernelMappingScheme::IndexingOrder indexing_order = [&]() {
+    if (reduction_dimensions.is_row_reduction &&
+        // Only try to vectorize+coales memory access for row of even size.
```
"for rows"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,7 +3171,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+  KernelMappingScheme::IndexingOrder indexing_order = [&]() {
+    if (reduction_dimensions.is_row_reduction &&
+        // Only try to vectorize+coales memory access for row of even size.
+        // For odd row size, every other row isn't aligned, so can't be
```
"For odd row sizes", "so it can't"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3133,11 +3204,23 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+  if (indexing_order == kLinearStridedIndexingX) {
+    if (reduction_dimensions.dimensions[2] % 2 == 0 &&
+        // Assuming XLA will perform the unrolling and LLVM will vectorize,
+        // disable the unroll for case that LLVM doesn't vectorize.
```
"for the case"

cheshire commented on tensorflow/compiler/xla/service/gpu/kernel_mapping_scheme.h:
```diff
@@ -133,12 +148,17 @@ class KernelMappingScheme {
+  // When num_threads_x threads process a total of tile_size_x
+  // elements in the X dimension of a tile, each threads process
+  // n=tile_size_x/num_threads_x elements.
+  // indexing_order_ define which tile's elements each thread reads.
```
"`indexing_order`", "defines", no separating newline.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/reduction_vectorization_test.cc:
```diff
@@ -0,0 +1,234 @@
+
+class ReductionVectorizationTest : public GpuCodegenTest {
+ protected:
+  void EnsureDeterminism(absl::string_view hlo_text) {
```
Why do you need this method?

nouiz committed:
Small change following code review.

nouiz reviewed (commented):
I updated the code from the comments.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1892,6 +1896,10 @@ bool MayPreventVectorization(const HloInstruction& hlo) {
         return false;
     }
+  } else if (hlo.opcode() == HloOpcode::kReduce) {
+    // TODO: check if the to_apply() attribute contain instruction
```
"contains"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1928,60 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+// are consecutive. If check_x_tile_bounds is true, then it will check
+// if the element index is in bound compared to tile_width before
+// calling emit_elem_function.
+static void Unroll(
```
We need a more self-descriptive name than `Unroll`, perhaps `UnrollInnerTileLoopForVectorization`?

nouiz replied:
I renamed it to UnrollInnerTileLoop, it doesn't always vectorize, it is an option. I find strange to specify an option in the name.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1928,60 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+  return b->CreateMul(thread_id_x, constant(x_num_steps));
+}
+
+// This function calls emit_elem_function() x_num_steps times.  If
```
Just `calls`, using backticks for function and variable names would make this more readable.

cheshire reviewed (changes requested):
Thanks, just only a few minor nits left, I think I'll be able to merge after the next round.

nouiz committed:
NFC following code review. comment typo and rename unroll to unrollInnerTileLoop.

cheshire reviewed (changes requested):
On internal benchmarks, I could see some slowdowns and no gains =(

Could you post the results from some of your benchmarks? Also results from micro-benchmarks (isolated runs from `run_hlo_module` with `--xla_hlo_profile`?)

nouiz commented:
Can you share one HLO graph that regress with this?
Also, on which GPU do you saw regression? I didn't benchmark on many GPUs or on older GPU generations. Here is an HLO that give me speed up the fused reduction in the optimized graph:

```
HloModule cluster_0__XlaCompiledKernel_true__XlaNumConstantArgs_0__XlaNumResourceArgs_0_.25

%max_ {
  %x = f16[] parameter(0)
  %y = f16[] parameter(1)
  ROOT %maximum = f16[] maximum(%x, %y)
}

%add_ {
  %x = f32[] parameter(0)
  %y = f32[] parameter(1)
  ROOT %add = f32[] add(%x, %y)
}

ENTRY %main {
  %arg0.1 = f16[32,131070]{1,0} parameter(0)
  %constant.3 = f16[] constant(-inf)
  %pad1 = f16[32,131070] pad(%arg0.1, %constant.3), padding=0_0_0x0_0_0
  %reduce.8 = f16[32]{0} reduce(%pad1, %constant.3), dimensions={1}, to_apply=%max_
  %broadcast.9 = f16[32,131070]{1,0} broadcast(%reduce.8), dimensions={0}
  %subtract.131070 = f16[32,131070]{1,0} subtract(%arg0.1, %broadcast.9)
  %exponential.11 = f16[32,131070]{1,0} exponential(%subtract.131070)
  %convert.12 = f32[32,131070]{1,0} convert(%exponential.11)
  %constant.13 = f32[] constant(0)
  %pad2 = f32[32,131070] pad(%convert.12, %constant.13), padding=0_0_0x0_0_0
  %reduce.18 = f32[32]{0} reduce(%pad2, %constant.13), dimensions={1}, to_apply=%add_
  %convert.19 = f16[32]{0} convert(%reduce.18)
  %broadcast.20 = f16[32,131070]{1,0} broadcast(%convert.19), dimensions={0}
  ROOT %divide.21 = f16[32,131070]{1,0} divide(%exponential.11, f16[32,131070]{1,0} %broadcast.20)
}
```

Here is the nvprof output:

```
==> OUT.softmaxf16_32x131070_131070._nvprof_TFV=old_master_xla=1.txt <==
2020-03-12 13:04:20.388311: E tensorflow/compiler/xla/tools/replay_computation.cc:476] iteration complete.
==9320== Profiling application: .../tensorflow-source.copy//bazel-out/k8-opt/bin/tensorflow/compiler/xla/tools/replay_computation_gpu softmax_template.txt.tmp --print_result=false --use_fake_data --num_runs=5
==9320== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   90.81%  4.8267ms         1  4.8267ms  4.8267ms  4.8267ms  [CUDA memcpy DtoH]
                    3.75%  199.14us         5  39.827us  39.328us  40.607us  fusion_1
                    2.77%  147.36us         5  29.472us  29.024us  29.792us  fusion
                    2.08%  110.56us         5  22.112us  21.600us  22.432us  reduce_8
                    0.34%  17.856us         1  17.856us  17.856us  17.856us  broadcast_2
                    0.25%  13.536us        10  1.3530us  1.0880us  1.8560us  [CUDA memset]

==> OUT.softmaxf16_32x131070_131070._nvprof_TFV=rb2_xla=1.txt <==
2020-03-12 12:57:39.328761: E tensorflow/compiler/xla/tools/replay_computation.cc:476] iteration complete.
==8924== Profiling application: .../tensorflow-source.copy//bazel-out/k8-opt/bin/tensorflow/compiler/xla/tools/replay_computation_gpu softmax_template.txt.tmp --print_result=false --use_fake_data --num_runs=5
==8924== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   94.13%  7.3283ms         1  7.3283ms  7.3283ms  7.3283ms  [CUDA memcpy DtoH]
                    2.17%  169.22us         5  33.843us  32.608us  35.040us  fusion_1
                    1.88%  146.72us         5  29.343us  28.896us  29.791us  fusion
                    1.42%  110.24us         5  22.047us  21.760us  22.176us  reduce_8
                    0.23%  17.792us         1  17.792us  17.792us  17.792us  broadcast_2
                    0.17%  13.312us        10  1.3310us  1.0560us  1.8880us  [CUDA memset]
```

nouiz committed:
Enable more vectorization by LLVMby generating simpler indexing. LLVM doesn't simplify correctly the complicated indexing generated. This souldn't affect the speed as ptxas was vectorizing the missing vectorization.

nouiz committed:
Vectorize reduction based on arch and dtypes.

nouiz commented:
I added 4 commits:
- The first one trigger some of the missing vectorization by LLVM. ptxas was vectorizing those already, but at least, now they are tested here.
- The 3rd one fix a regression. We where using the columns reduction indexing in some cases.
- The 2nd and 4th limit the vectorization on P100 GPUs. We now only vectorize when the tile fit exactly and for dtype of up to 32 bits. 

On P100, with float16, I see 1.23x and 2.33x speed up. In float32, I see 1.08x and 1.10x speed up.

The PR can be benchmarked. But I still need to update the tests to have expect the right vectorization depending of the hardware. I need to find how to detect the current GPU in the test.

nouiz committed:
Only trigger reduction columns indexing for columns reductions.

nouiz committed:
For row reduction do not try to vectorize on P100 when the tile doesn't fit perfectly and for dtype bigger then 32 bits.

nouiz committed:
Update test to check the GPU arch for the expected ptx.

nouiz committed:
clang-format

nouiz committed:
Crash fix. We where not returning the right reduction tile size.

nouiz committed:
Small refactoring. Instead of recomputing a value, store it. This is safer as all the numbers are computed and validated once.

nouiz commented:
I updated the tests to expect the right PTX depending of the GPU arch.
I also fixed crashed that I introduced today by an earlier in d9ae69f04ba944384d117662b888c43cb7e0bf72

I did a small refactoring to help that kind of error to not pop up again here.

cheshire commented:
@nouiz I'm a bit confused by your benchmark, do you mean the runtime reduction in `fusion_1` from `39us` to `33us`?

nouiz commented:
exact

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -2035,6 +2099,16 @@ static IrArray::Index GetUnnormalizedIndex(
   DCHECK_EQ(normalized_shape_index.size(), 3);
+  // If the normalization only add a new dimensions of size 1,
+  // generate simpler indexing. LLVM doesn't always simplify the more
+  // complicated indexing and this prevent him from vectorizing some
```
"prevents it"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -2035,6 +2099,16 @@ static IrArray::Index GetUnnormalizedIndex(
+  if (unnormalized_shape.rank() == 2) {
+    DCHECK_EQ(normalized_shape_index.dims()[0], 0);
+    auto multidim = normalized_shape_index.multidim();
+    return IrArray::Index({multidim[1], multidim[2]}, unnormalized_shape,
```
Sorry, I am a bit confused on what is going on here, this code is new.
Why rank 2? Isn't `[2]` out of bounds in this case?

nouiz replied:
Their is a "normalization" step of the input shapes to a 3d shape. So a matrix of shape MxN is converted to 1xMxN. A 3d tensor of shape MxNxK that we reduce on the first 2 dimensions is converted to a shape of 1xM*NxK, ...
This help simplify the code generator as the shapes he do not need to optimize those part.

But when we want to trigger the read, we need to "unnormalize" the shapes/indexing. This unnormalization isn't always well cleaned up by LLVM. This unnormalization is also done for each elements read, not once per threads.

So here, I simplify remove this unnormalization step when it isn't needed. This clean up the not optimized LLVM code and allow LLVM to vectorize more this case.

The unnormalization is taking the normalized 3d indexing, converting it to a 1d indexing, then converting it back to the original indexing. It is a afair amount of LLVM code.

as multidim is the normalized shapes that is always 3d, multidim[2] is valid.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3133,12 +3188,54 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+  int num_partial_results = 1;
+  KernelMappingScheme::IndexingOrder indexing_order = [&]() {
+    if (reduction_dimensions.is_row_reduction &&
+        // P100, only ttry to vectorize+coales memory access when the
```
"try"

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3133,12 +3188,54 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
+  KernelMappingScheme::IndexingOrder indexing_order = [&]() {
+    if (reduction_dimensions.is_row_reduction &&
+        // P100, only ttry to vectorize+coales memory access when the
+        // tile size fit exactly and dtypes <= 32 bits
```
"fits"

cheshire reviewed (changes requested):
Some nits for the new code.

Is it possible to add the microbenchmark results you've posted in the commit description? Also maybe it's possible to have a smaller benchmark, I'm not sure why fusion is needed to demonstrate the performance gain?

nouiz committed:
Fix comment typo

nouiz reviewed (commented):
I added a the benchmark information in an hopefully clearer format in the description.
As the benchmark is the softmax operation, I think it is a realistic graph. As we mostly see the speed up in the fused group, I would have a tendency to keep it as it.

We see more speed up in the fused reduction because the loads aren't consecutive in the instructions. This make a difference in the hardware.

The unrolled computation is handled like this:
```
load
compute
load
compute
...
```

I have implemented locally changes that change the order to be:
```
load
load
compute
compute
load
load
compute
compute
...
```

And then I also implemented manual vectorisation. This was to support odd row size. But the speed up isn't great and it is a significant amount of code. So I do not plan to upstream all of it.

But if you think that up-streaming the new order of operations and the manual vectorisation code is useful, I can do it. The emitted vectorized code looks like this:
```
load 2xdtype
compute
compute
load 2xdtype
compute
compute
...
```

cheshire reviewed (dismissed):
Thanks!

nouiz commented:
Their is a CI failure. I didn't see it before. It happens non-deterministically, but I'm able to reproduce it from time to time. I'll look into it .

nouiz committed:
Do a simple unnormalize only for row major memory format and exact shape.

nouiz commented:
Fixed. I needed to check that the layout is row major for one of the simplification.

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -2102,8 +2102,11 @@ static IrArray::Index GetUnnormalizedIndex(
+      unnormalized_shape.dimensions()[0] == normalized_shape_index.dims()[1] &&
+      unnormalized_shape.dimensions()[1] == normalized_shape_index.dims()[2] &&
+      unnormalized_shape.layout().minor_to_major(1) == 0) {
     DCHECK_EQ(normalized_shape_index.dims()[0], 0);
```
This DCHECK is wrong, it should check that the trivial dimension 0 has value 1.
DCHECK_EQ(normalized_shape_index.dims()[0], 1)

cheshire replied:
Going further, can we make it a CHECK_EQ instead?

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1978,32 +2035,39 @@ void IrEmitterUnnested::EmitTile(
+      /*step=*/constant(1), [&](llvm::Value* y_indvar) {
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unrollInnerTileLoop = [&](bool check_x_tile_bounds) {
```
this should be named unroll_inner_tile_loop according to the style guide. It is still a variable.

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1928,60 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+      llvm::Value* x_loc = b_.CreateAdd(constant(j * step_x * vector_size + i),
+                                        start_offset_x, "x_loc");
+      IrArray::Index source_idx_x =
+          source_idx.AddOffsetToDim(y_loc, kDimY, &b_)
```
You could pull this line out of this function and already pass a source_idx_y to the function.
Less generated code, although it would certainly be optimized away by LLVM anyway.

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3116,11 +3179,6 @@ ReductionCodegenInfo IrEmitterUnnested::ComputeReductionCodegenInfo(
       !IsUnrollingColumnReductionBeneficial(unnested_hlo, input_shape,
                                             reduction_dimensions.dimensions[2]);
 
-  if (!dilated_x && !reduction_dimensions.is_row_reduction) {
```
This change makes the variable dilated_x unused, so it should be deleted.

akuegel reviewed (commented):
Unfortunately this PR had to be rolled back, because it is failing a DCHECK when running the test tensorflow/compiler/xla/client/lib:arithmetic_test_gpu without -c opt (i.e. with NDEBUG enabled).
Check failed: shape.IsArray() (s32[3], s32[3])
While at it, I added a few more comments.

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1928,60 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+  auto constant = [&](int64 val) {
+    return llvm::ConstantInt::get(index_ty, val);
+  };
+  for (int j = 0; j < x_num_steps / vector_size; j++) {
```
The int's should be int64's

akuegel commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -1920,13 +1928,60 @@ static llvm::Value* GetStartOffsetX(const KernelMappingScheme& mapping_scheme,
+    bool check_x_tile_bounds, int64 x_num_steps, int64 step_x,
+    int64 vector_size, const string& loop_name, KernelSupportLibrary* ksl,
+    llvm::Value* start_offset_x, llvm::Value* y_loc, llvm::Value* tile_width,
+    IrArray::Index& source_idx, llvm::IRBuilder<>& b_,
```
Please pass source_idx as const reference (it is not changed), and a pointer IrBuilder instead of a reference (this is according to the style guide).

nouiz commented:
New PR with the CHECK fix and the other small code changes:
https://github.com/tensorflow/tensorflow/pull/38136
